aip/general/0128.md

```diff
@@ -103,10 +103,9 @@ is a comprehensive reference to declarative-friendly guidance in other AIPs:
 
 - Resources **must** use user-settable resource IDs: see AIP-133.
 - Resources **must** permit "create or update": see AIP-134.
-- Resources **should** permit "delete if existing": see AIP-135.
 - Resources **should not** employ custom methods: see AIP-136.
 - Resources **must** use the `Update` method for repeated fields: see AIP-144.
 - Resources **must** include certain standard fields: see AIP-148.
 - Resources **must** have an `etag` field: see AIP-154.
-- Resources **must** provide change validation: see AIP-163.
-- Resources **should** support soft delete: see AIP-164.
+- Resources **should** provide change validation: see AIP-163.
+- If a resource cannot be recreated with a previously used id, and it is feasible to support soft-delete, it **must** support soft delete: see AIP-164.
```

Comment:
Soft delete is not used by declarative tools, is an inconvenience to them actually, and is complicated to implement. I'd rather not recommend it.

Reply:
Are you specifically thinking that undelete must be supported? AIP 164 doesn't require that.

Reply:
I should probably extrapolate on my thoughts a bit here. I'll make sure to update the commit message as well.

It's true that for GCP, declarative clients like Terraform do not use soft delete today. But if the resource supports an undelete scenario to enable the re-use of ids, I kind of feel like Terraform should support that.

Pros:

- enables resources to re-use resource IDs that were inaccessible otherwise.

Cons

- potentially exposes state that the user may not have had access to previously (e.g. for a database user-data). Although if the user has the permissions to create the resource with that ID in the first place, they should have access to that data.

The safe thing would be to not use soft-delete in declarative clients at the expense of imperative overhead when undeleting (calling a custom method, importing the resource into the declarative client). But that is a slightly poorer UX for customers who want to or have to do everything declaratively. 

Thoughts?

Reply:
How about should not implement soft delete unless it's not possible to reuse the id (e.g., to recreate from a backup or snapshot), and otherwise must implement soft delete and undelete. AIP 164 doesn't require undelete.

Reply:
done! sounds good.